### PR TITLE
CRM-21401 Fix issues with groupBy on Maria10.2

### PR DESCRIPTION
**Pull request by seamuslee001**
Overview
----------------------------------------
This aims to fix some groupBy issues on Maria10.2 mysql engine. 

@mattwire @kcristiano I believe this should fix your issues also ping @totten @monishdeb

---

 * [CRM-21401: Fix issues running CiviCRM on Maria 10.2](https://issues.civicrm.org/jira/browse/CRM-21401)

### CODE REVIEW

File: CRM/Contact/BAO/Query.php [hunk 1]
```diff
@@ -4749,7 +4748,7 @@ public static function getGroupByFromSelectColumns($selectClauses, $groupBy = NU
           $primaryKey = "{$tableName}.id";
           // exclude columns which are already included in groupBy and aggregate functions from select
           // CRM-18439 - Also exclude the columns which are functionally dependent on columns in $groupBy (MySQL 5.7+)
-          if (!in_array($selectColumn, $groupBy) && !in_array($primaryKey, $groupBy) && preg_match($regexToExclude, trim($selectColumn)) !== 1) {
```
Comment (seamuslee001) [inline CRM/Contact/BAO/Query.php]: Even tho we do group by on primary key need to group by on other columns as well for full_group_by to work

File: CRM/Contact/BAO/Query.php [hunk 2]
```diff
@@ -4724,7 +4724,6 @@ public static function getGroupByFromOrderBy(&$groupBy, $orderBys) {
    */
   public static function getGroupByFromSelectColumns($selectClauses, $groupBy = NULL) {
     $groupBy = (array) $groupBy;
-    $mysqlVersion = CRM_Core_DAO::singleValueQuery('SELECT VERSION()');
```
Comment (seamuslee001) [inline CRM/Contact/BAO/Query.php]: This is just redundant

**Comment by mattwire** [comment]
Here's another:
`Nov 07 10:11:29  [info] $Fatal Error Details = Array
(
    [callback] => Array
        (
            [0] => CRM_Core_Error
            [1] => handle
        )

    [code] => -1
    [message] => DB Error: unknown error
    [mode] => 16
    [debug_info] =>                                                                                                                               
SELECT contact_id,                                                                                                                          
       civicrm_membership_type.name,                                                                                                                  
       start_date,
       civicrm_membership_type.minimum_fee 
FROM   civicrm_membership  
       LEFT JOIN civicrm_membership_type ON civicrm_membership.membership_type_id=civicrm_membership_type.id

WHERE  contact_id IN ( 17,41,73,86,87,88,119,151,157,162,174,198,199,203,203,203,203,203,203,203 )
	AND civicrm_membership.membership_type_id IN (9,10,11,12)
AND    is_test = 0                                                                                                                          
GROUP BY contact_id                                                                                                                         
 [nativecode=1055 ** 'drupalclea_hbege.civicrm_membership_type.name' isn't in GROUP BY]
    [type] => DB_Error
    [user_info] =>                                                                                                                               
SELECT contact_id,                                                                                                                          
       civicrm_membership_type.name,                                                                                                                  
       start_date,
       civicrm_membership_type.minimum_fee 
FROM   civicrm_membership  
       LEFT JOIN civicrm_membership_type ON civicrm_membership.membership_type_id=civicrm_membership_type.id

WHERE  contact_id IN ( 17,41,73,86,87,88,119,151,157,162,174,198,199,203,203,203,203,203,203,203 )
	AND civicrm_membership.membership_type_id IN (9,10,11,12)
AND    is_test = 0                                                                                                                          
GROUP BY contact_id                                                                                                                         
 [nativecode=1055 ** 'drupalclea_hbege.civicrm_membership_type.name' isn't in GROUP BY]
    [to_string] => [db_error: message="DB Error: unknown error" code=-1 mode=callback callback=CRM_Core_Error::handle prefix="" info="                                                                                                                              
SELECT contact_id,                                                                                                                          
       civicrm_membership_type.name,                                                                                                                  
       start_date,
       civicrm_membership_type.minimum_fee 
FROM   civicrm_membership  
       LEFT JOIN civicrm_membership_type ON civicrm_membership.membership_type_id=civicrm_membership_type.id

WHERE  contact_id IN ( 17,41,73,86,87,88,119,151,157,162,174,198,199,203,203,203,203,203,203,203 )
	AND civicrm_membership.membership_type_id IN (9,10,11,12)
AND    is_test = 0                                                                                                                          
GROUP BY contact_id                                                                                                                         
 [nativecode=1055 ** 'drupalclea_hbege.civicrm_membership_type.name' isn't in GROUP BY]"]
)
`

**Comment by mattwire** [comment]
@seamuslee001 Looks like this fixes the two issues I found yesterday.  Getting lot's of breakage with extensions on mariadb 10.2 due to only_full_group_by issues :-(

**Comment by seamuslee001** [comment]
@mattwire interesting i have production running fine on Maria 10.1 with only full group by something certainly up with Maria10.2 i think

**Comment by mattwire** [comment]
@seamuslee001 Sorry, getting lot's of issues with this version: `mariadb  Ver 15.1 Distrib 10.1.26-MariaDB, for debian-linux-gnu (x86_64) using readline 5.2`

Though this patch fixes two of them :-)  Saw some chatter on mattermost about different versions of mariadb having different default config.

**Comment by kcristiano** [comment]
@seamuslee001 I am still having issues with mariadb 10.2.10  None of these dashboards will open contribution, membership, events.    I can test further when I get a chance.  Same issues are in 4.7.28-rc.   4.7.27 is fine.   I'll look at my mariadb config on 10.2 and see if it differes from my 10.1 setup.  I am also using 10.1 for production and 10.2 only fro dev/testing

**Comment by eileenmcnaughton** [comment]
@seamuslee001 (& @monishdeb ) I don't think I'm ever going to feel comfortable merging this because
1) we've been burnt a lot of full group by fixes
2) it includes changes to several deep functions in one PR
3) some of those lines are already associated with performance issues and
4) there are no tests
5) I don't really know how to verify that there are no resulting regressions

I think we need to figure out what we are trying to acheive for full group by - if we are trying to get to the point where it 'just works' on mysql with full group by I think we should
1) have a way of marking functions that are not yet full group y & simply turn the mode on & off before & after them 
2) if necessary explicitly do not support full group by if the mysql user does not have permission to turn it off & on
3) switch to normal CI being in full group by mode & prevent any new ones
4) as we touch lines of code that are not fullGroupByCompliant (per 1) we fix them, test them etc

**Comment by seamuslee001** [comment]
@eileenmcnaughton this has also been essentially dealt with by the fact that we have excluded MariaDB installs from having the full groupBy changes

**Comment by eileenmcnaughton** [comment]
@seamuslee001 so we can close?

**Comment by seamuslee001** [comment]
I think so and then we can look at the full group by stuff

**Pull request closed without merging**